### PR TITLE
*: fix mailbox leak

### PR DESCRIPTION
**Pull request by BusyJay**
### What problem does this PR solve?

Problem Summary:

#9019 shows that even after using LRU cache, a large memory are still held by router. This is because the cache is used everywhere, and waiting for all cache to be cleared can be very slow. This PR fixes the problem completely by discarding the memory when all receivers are dropped.

### Check List 

Tests 

- Unit test
- Integration test

### Release note 
- Fix memory leak by slow eviction of mailbox cache

### CODE REVIEW

File: components/cdc/src/endpoint.rs [hunk 1]
```diff
@@ -1203,8 +1204,8 @@ impl<T: 'static + RaftStoreRouter<RocksEngine>> Runnable for Endpoint<T> {
                 cb,
             } => {
                 debug!("downstream was initialized"; "downstream_id" => ?downstream_id);
-                downstream_state
-                    .compare_and_swap(DownstreamState::Uninitialized, DownstreamState::Normal);
+                let _ = downstream_state
```
Comment (BusyJay) [inline components/cdc/src/endpoint.rs:1207]: Is it safe to ignore error? @5kbpers @overvenus

File: components/cdc/src/endpoint.rs [hunk 2]
```diff
@@ -340,9 +341,9 @@ impl<T: 'static + RaftStoreRouter<RocksEngine>> Endpoint<T> {
                 if is_last {
                     let delegate = self.capture_regions.remove(&region_id).unwrap();
                     if let Some(reader) = self.store_meta.lock().unwrap().readers.get(&region_id) {
-                        reader
+                        let _ = reader
```
Comment (BusyJay) [inline components/cdc/src/endpoint.rs:344]: Another similar case @5kbpers

File: components/cdc/src/endpoint.rs (hunk 1 above)
Comment (5kbpers) [reply components/cdc/src/endpoint.rs:1207]: yes

Comment (BusyJay) [reply components/cdc/src/endpoint.rs:1214]: If so, why not just use `store`?

File: components/raftstore/src/store/fsm/store.rs [hunk 3]
```diff
@@ -1324,6 +1324,8 @@ impl<EK: KvEngine, ER: RaftEngine> RaftBatchSystem<EK, ER> {
             warn!("set thread priority for raftstore failed"; "error" => ?e);
         }
         self.workers = Some(workers);
+        // This router will not be accessed again, free all caches.
+        self.router.clear_cache();
```
Comment (Connor1996) [inline components/raftstore/src/store/fsm/store.rs:1328]: Seems it's the main changes to fix the leak. Does the crossbeam update related to this fix?

Comment (BusyJay) [reply components/raftstore/src/store/fsm/store.rs:1328]: See https://github.com/crossbeam-rs/crossbeam/pull/669

File: components/cdc/src/endpoint.rs (hunk 1 above)
Comment (overvenus) [reply components/cdc/src/endpoint.rs:1214]: Could you unwrap it?

**Review by sticnarf — APPROVED** [review]
LGTM

File: components/cdc/src/endpoint.rs (hunk 1 above)
Comment (5kbpers) [reply components/cdc/src/endpoint.rs:1214]: Because it could be `Stopped` state

File: components/raftstore/src/store/fsm/store.rs (hunk 3 above)
Comment (gengliqi) [reply components/raftstore/src/store/fsm/store.rs:1332]: Seems this PR has not been included in the release version of crossbeam yet?

File: components/cdc/src/endpoint.rs (hunk 1 above)
Comment (overvenus) [reply components/cdc/src/endpoint.rs:1214]: Is it safe to transit state from `Stopped` to `Normal`?

File: components/concurrency_manager/Cargo.toml [hunk 4]
```diff
@@ -21,8 +21,8 @@ fail = "0.4"
 
 # FIXME: switch to the crates.io version after crossbeam-skiplist is released
```
[the remark this answers is not in the document]
Comment (BusyJay) [reply components/concurrency_manager/Cargo.toml:22]: Actually we still want to switch to the released version later as two patches are all merged.

File: components/raftstore/src/store/fsm/store.rs (hunk 3 above)
Comment (BusyJay) [reply components/raftstore/src/store/fsm/store.rs:1332]: Yes, that's the reason why a forked repository is used.

File: components/cdc/src/endpoint.rs (hunk 1 above)
Comment (5kbpers) [reply components/cdc/src/endpoint.rs:1215]: No, so we should not unwrap it here

File: components/raftstore/src/store/fsm/store.rs (hunk 3 above)
Comment (NingLin-P) [reply components/raftstore/src/store/fsm/store.rs:1332]: `self.router`'s cache only contains 1024 slots, the main fix should be the crossbeam patch?

Comment (BusyJay) [reply components/raftstore/src/store/fsm/store.rs:1332]: Yes.

**Review by NingLin-P — APPROVED** [review]
LGTM

**Review by Connor1996 — APPROVED** [review]
LGTM

**Review by gengliqi — APPROVED** [review]

**Comment by hi-rustin** [comment]
> /merge

You only need to trigger /merge once, and if the CI test fails, you just re-trigger the test that failed and the bot will merge the PR for you after the CI passes.

**Pull request merged**